adapters/insticator/insticator.go

```diff
@@ -0,0 +1,526 @@
+	if imp.Banner == nil && imp.Video == nil {
+		return openrtb2.Imp{}, &errortypes.BadInput{
+			Message: fmt.Sprintf("Imp ID %s must have at least one of [Banner, Video] defined", imp.ID),
+		}
+	}
+
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return openrtb2.Imp{}, &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+
+	var insticatorExt openrtb_ext.ExtImpInsticator
+	if err := json.Unmarshal(bidderExt.Bidder, &insticatorExt); err != nil {
+		return openrtb2.Imp{}, &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+
+	var impExt Ext
+	impExt.Insticator.AdUnitId = insticatorExt.AdUnitId
+	impExt.Insticator.PublisherId = insticatorExt.PublisherId
+
+	impExtJSON, err := json.Marshal(impExt)
+	if err != nil {
+		return openrtb2.Imp{}, &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+
+	imp.Ext = impExtJSON
+
+	// Validate Video if it exists
+	if imp.Video != nil {
+		videoCopy, err := validateVideoParams(imp.Video)
+
+		imp.Video = videoCopy
+
+		if err != nil {
+			return openrtb2.Imp{}, &errortypes.BadInput{
+				Message: err.Error(),
+			}
+		}
+	}
+	return imp, nil
+}
+
+func validateVideoParams(video *openrtb2.Video) (*openrtb2.Video, error) {
+	videoCopy := *video
+	if (videoCopy.W == nil || *videoCopy.W == 0) ||
+		(videoCopy.H == nil || *videoCopy.H == 0) ||
+		videoCopy.MIMEs == nil {
+
+		return nil, &errortypes.BadInput{
+			Message: "One or more invalid or missing video field(s) w, h, mimes",
+		}
+	}
+
+	// Validate optional parameters and remove invalid ones
+	cleanedVideo, err := validateOptionalVideoParams(&videoCopy)
+	if err != nil {
+		return nil, err
+	}
+
+	return cleanedVideo, nil
+}
+
+func makeReqExt(request *openrtb2.BidRequest) ([]byte, error) {
+	var reqExt reqExt
+
+	if len(request.Ext) > 0 {
+		if err := json.Unmarshal(request.Ext, &reqExt); err != nil {
+			return nil, err
+		}
+	}
+
+	if reqExt.Insticator == nil {
+		reqExt.Insticator = &reqInsticatorExt{}
+	}
+
+	if reqExt.Insticator.Caller == nil {
+		reqExt.Insticator.Caller = make([]InsticatorCaller, 0)
+	}
+
+	reqExt.Insticator.Caller = append(reqExt.Insticator.Caller, CALLER)
+
+	return json.Marshal(reqExt)
+}
+
+func resolveBidFloor(bidFloor float64, bidFloorCur string, reqInfo *adapters.ExtraRequestInfo) (float64, error) {
+	if bidFloor > 0 && bidFloorCur != "" && strings.ToUpper(bidFloorCur) != "USD" {
+		floor, err := reqInfo.ConvertCurrency(bidFloor, bidFloorCur, "USD")
+		return roundTo4Decimals(floor), err
+	}
+
+	return bidFloor, nil
+}
+
+// roundTo4Decimals function
+func roundTo4Decimals(amount float64) float64 {
+	return math.Round(amount*10000) / 10000
+}
+
+func validateOptionalVideoParams(video *openrtb2.Video) (*openrtb2.Video, error) {
+	v := reflect.ValueOf(video).Elem()
+	t := v.Type()
+
+	for i := 0; i < t.NumField(); i++ {
+		field := t.Field(i)
+		fieldValue := v.Field(i)
+
+		// Convert the field name to camelCase for matching in optionalVideoParams map
```

Comment:
Usage of reflection should be avoided.  

Also looks like `validateOptionalVideoParams` is validating if correct values are parsed and assigned to the fields of `*openrtb2.Video` struct. This step is not needed.  

Because adapter code will receive parsed values. Here `MakeRequests` receives `request *openrtb2.BidRequest`.  and `request.Imp[$].video` is already validated during parsing.

```
func (a *adapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) 
```
Parsing done before MakeRequests call ensures that incoming values are passable for struct